Doc: jobs option and package build setting (#496)
python-wheel-build/fromager: docs/how-to.md

# How To

> ## Use --jobs option, new cpu_cores_per_job and memory_per_job_gb settings

Author: okay, changed it

The `--jobs` option of fromager allows to set the maximum number of wheel build jobs to run in parallel. Below is an example which uses the `--jobs` option along with the bootstrap command

> `fromager --jobs 4 bootstrap torch`

Author: Added the info

The `cpu_cores_per_job` is a package setting that allows to scale parallel jobs by available CPU cores. The default value is set to 1 which indicates as many parallel jobs as CPU logical cores.
Example: `cpu_cores_per_job = 2` indicates allocating 2 cores per job
This setting should always have value greater than or equal to 1

The `memory_per_job_gb` is a package setting that allows to scale parallel jobs by available virtual memory without swap. The default value is set to 1.0 which indicates that each parallel job requires 1 GB virtual memory
Example: `memory_per_job_gb = 0.5` indicates that each parallel job requires 512 MB virtual memory

> This setting should always have value greater than or equal to 0.1

Author: I have tried to add information about this but am not sure about it

`fromager --jobs 4 bootstrap torch`

For this example, the maximum number of jobs fromager will run in parallel is 4.

The `cpu_cores_per_job` is a package setting that allows to scale parallel jobs by available CPU cores. The default value is set to 1 which indicates as many parallel jobs as CPU logical cores.
Example: `cpu_cores_per_job = 2` indicates allocating 2 cores per job
This setting should always have value greater than or equal to 1

The `memory_per_job_gb` is a package setting that allows to scale parallel jobs by available virtual memory without swap. The default value is set to 1.0 which indicates that each parallel job requires 1 GB virtual memory
Example: `memory_per_job_gb = 0.5` indicates that each parallel job requires 512 MB virtual memory
This setting should always have value greater than or equal to 0.1

The above mentioned options are related to build systems and can be used as a per package setting when multiple cores and significant amount of memory is available on a build system. On the other hand, the `--jobs` option can be

> used to spin up paralell jobs to improve speed of the wheel build process.

Reviewer: I think `--jobs` overrides the default calculations based on the other settings, doesn't it? So if someone passes `--jobs 4` then at most 4 processes will run in parallel when building a given wheel, but if they don't use that option fromager computes a number of jobs and that value can be influenced based on the per-package settings.

And the jobs are all within the context of building a single wheel. We are not currently building multiple wheels in parallel, but some of the expensive builds do run the C++ compiler in parallel.

Author: Oh right. I have made changes and pushed latest content

`fromager --jobs 4 bootstrap torch`

For this example, the maximum number of jobs fromager will run in parallel is 4.

The `cpu_cores_per_job` is a package setting that allows to scale parallel jobs by available CPU cores. The default value is set to 1 which indicates as many parallel jobs as CPU logical cores.
Example: `cpu_cores_per_job = 2` indicates allocating 2 cores per job
This setting should always have value greater than or equal to 1

The `memory_per_job_gb` is a package setting that allows to scale parallel jobs by available virtual memory without swap. The default value is set to 1.0 which indicates that each parallel job requires 1 GB virtual memory
Example: `memory_per_job_gb = 0.5` indicates that each parallel job requires 512 MB virtual memory
This setting should always have value greater than or equal to 0.1

The above mentioned options are related to build systems and can be used as a per package setting when multiple cores and significant amount of memory is available on a build system.

> On the other hand, the `--jobs` overrides the default calculations based on the other settings. For example, when someone passes `--jobs 4` then at most 4 processes will run in parallel when building a given wheel, but if they don't use this option, fromager computes a number of jobs and that value can be influenced based on the per-package settings.

Reviewer: ```suggestion
On the other hand, the `--jobs` overrides the default calculations based on the other settings. For example, when you pass `--jobs 4` then at most 4 processes will run in parallel when building a given wheel. By default, fromager computes a number of jobs and that value can be influenced based on the per-package settings.
```

Author: Okay, done

> Note that the jobs are all within the context of building a single wheel.

Reviewer: OK, this information is right, but I suggest moving it up to the top of the section (before what is now line 5) as an introduction.

Author: Okay, done
